### PR TITLE
Support CMAKE_OSX_ARCHITECTURES

### PR DESCRIPTION
**Pull request by alexreinking**
As requested at the last Halide dev meeting, I have drafted an implementation of support for CMAKE_OSX_ARCHITECTURES. This is again quite complicated, in part because I attempted to avoid breaking existing user code.

As it stands, multi-target **alternative selection** and **feature specification** are stepping on each others' toes.

I'll write more tomorrow.

---

Breaking suggestion:

* The `TARGETS` argument and the `Halide_TARGET` variable shall take only the following values:
  * `host` (passed verbatim to the generator, but resolved for feature selection)
  * `cmake`
  * `ARCH-BITS-OS` (a triple)
* When `CMAKE_OSX_ARCHITECTURES` is set, `cmake` expands to a list of compatible architectures, e.g. `arm-64-osx;x86-64-osx`. This is reflected in `Halide_CMAKE_TARGET`. `host` expands to a single triple corresponding to the architecture of the `cmake` executable itself (this enables compatibility with the `arch` command).
* The `FEATURES` argument applies a list of features to ALL alternatives on ALL platforms (useful for, e.g. `user_context`, name mangling, etc.)
* The `FEATURES[platform]` arguments set up a list of alternatives (e.g. `FEATURES[x86-64-linux] "avx2-sse4" "sse4"`) for a particular triple.

What this doesn't resolve is how to do things like `TARGETS cmake-no_bounds_query cmake` effectively. Perhaps the following adjustments to the above help:

* Rename `FEATURES[platform]` to `MULTITARGET[platform]`
* Add `MULTITARGET` (unadorned) to provide the default value for unspecified `[platform]`s.

Specifying an empty entry in `FEATURES`/`MULTITARGET` or whatever is challenging in CMake because it really wants to drop empty function arguments. We might need to add a dummy feature like `none` (either in the actual Generator or just in the build system) that does nothing but isn't the empty string.

### CODE REVIEW

File: cmake/HalideGeneratorHelpers.cmake [hunk 1]
```diff
@@ -142,32 +144,223 @@ function(add_halide_generator TARGET)
...
 endfunction()
 
+function(_Halide_invoke_generator)
+    cmake_parse_arguments(
+        PARSE_ARGV 0 ARG ""
+        "BASE_NAME;FUNCTION_NAME;GENERATOR;GRADIENT_DESCENT;TYPE"
+        "COMMAND;DEPENDS;EXTRA_OUTPUTS;PARAMS;PLUGINS;TARGETS"
+    )
+
+    ## "hash table" of extra outputs to extensions
```
Comment (derek-gerstmann) [inline cmake/HalideGeneratorHelpers.cmake:177]: Can we add 'hlpipe' here too?  It should already be hooked up in the generator interface.

File: cmake/HalideGeneratorHelpers.cmake [hunk 2]
```diff
@@ -142,32 +144,223 @@ function(add_halide_generator TARGET)
...
+
+function(_Halide_lipo TARGET)
+    set(merged_libs ${ARGN})
+    list(TRANSFORM merged_libs REPLACE "^(.+)$" "$<TARGET_FILE:\\1>"
+         OUTPUT_VARIABLE merged_libs_files)
+    list(TRANSFORM merged_libs REPLACE "^(.+)$" "$<COMPILE_ONLY:\\1>"
+         OUTPUT_VARIABLE merged_libs_targets)
+
+    find_program(LIPO lipo REQUIRED)
+    add_custom_command(
```
Comment (derek-gerstmann) [inline cmake/HalideGeneratorHelpers.cmake:354]: Nicely done, and with all the dependencies setup too!

File: src/autoschedulers/adams2019/CMakeLists.txt [hunk 3]
```diff
@@ -22,57 +22,16 @@ target_link_libraries(adams2019_cost_model.generator PRIVATE Halide::Generator)
...
-                   TARGETS cmake)
+                   FEATURES[x86-64-osx] avx2 sse41
+                   FEATURES[arm-64-osx] arm_dot_prod-arm_fp16)
+
 add_halide_library(adams2019_train_cost_model FROM adams2019_cost_model.generator
                    GENERATOR train_cost_model
                    FUNCTION_NAME train_cost_model
-                   TARGETS cmake
+                   FEATURES[x86-64-osx] avx2 sse41
+                   FEATURES[arm-64-osx] arm_dot_prod-arm_fp16
```
Comment (derek-gerstmann) [inline src/autoschedulers/adams2019/CMakeLists.txt:32]: Are these features available on all Apple Silicon?

**Review by derek-gerstmann — APPROVED** [review]

File: cmake/HalideGeneratorHelpers.cmake (hunk 1 above)
Comment (alexreinking) [reply cmake/HalideGeneratorHelpers.cmake:177]: Definitely. Didn't know it was missing!

File: cmake/BundleStatic.cmake [hunk 4]
```diff
@@ -22,12 +22,10 @@ function(_bundle_static_check_output VAR)
...
 
-function(_bundle_static_is_apple_libtool result item)
+function(_bundle_static_is_apple_libtool result_var item)
     _bundle_static_check_output(version_info "${item}" -V)
-    if (version_info MATCHES "Apple, Inc.")
-        set(result 1 PARENT_SCOPE)
-    else ()
-        set(result 0 PARENT_SCOPE)
+    if (NOT version_info MATCHES "Apple, Inc.")
+        set(${result_var} 0 PARENT_SCOPE)
```
Comment (alexreinking) [inline cmake/BundleStatic.cmake:28]: This is a drive-by fix.

**Comment by alexreinking** [comment]
Topics for discussion at the dev meeting:

1. How to test this on CI?
2. How to test build sanity on iOS?

**Comment by alexreinking** [comment]
Finally, we have a clean build! The `halide-asan` test failure is due to an LLVM build bug that was fixed in [a revert five days ago](https://github.com/llvm/llvm-project/commit/a15de177728ec83f37bdd8f39d4a8f57e95c3d21).

**Comment by derek-gerstmann** [comment]
Testing on my MacBook Pro M1 Max and everything appears to work!  I agree, it'd be nice to find a testing strategy for both desktop and mobile.

**Comment by alexreinking** [comment]
from dev meeting:

- Need to collect multiple extra-outputs in the output variables (how to test?)
- Need to provide a way to generate the c sources as a byproduct, rather than going through the configured compiler

**Comment by alexreinking** [comment]
This is ready on my end. @steven-johnson -- we should coordinate adding testing to the buildbots. Should we wait to land this until we have that?

**Comment by steven-johnson** [comment]
> This is ready on my end. @steven-johnson -- we should coordinate adding testing to the buildbots. Should we wait to land this until we have that?

yes, probably

**Review by steven-johnson — APPROVED** [review]
Is there documentation/README for this anywhere...?

**Comment by alexreinking** [comment]
> Is there documentation/README for this anywhere...?

No, not yet

**Comment by alexreinking** [comment]
Landing this with the caveat that it might need adjustment to support IOS cross-compilation via Xcode. At the moment it seems to have not broken existing workflows and has (locally) enabled universal builds on macOS.
